presto-hudi/src/main/java/com/facebook/presto/hudi/HudiPageSource.java

```diff
@@ -0,0 +1,243 @@
+import static java.lang.Double.parseDouble;
+import static java.lang.Float.floatToRawIntBits;
+import static java.lang.Float.parseFloat;
+import static java.lang.Long.parseLong;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class HudiPageSource
+        implements ConnectorPageSource
+{
+    private final Block[] prefilledBlocks;
+    private final int[] delegateIndexes;
+    private final ConnectorPageSource delegate;
+
+    public HudiPageSource(
+            List<HudiColumnHandle> columns,
+            Map<String, String> partitionKeys,
+            ConnectorPageSource delegate,
+            TimeZoneKey timeZoneKey,
+            TypeManager typeManager)
+    {
+        int size = requireNonNull(columns, "columns is null").size();
+        requireNonNull(partitionKeys, "partitionKeys is null");
+        this.delegate = requireNonNull(delegate, "delegate is null");
+
+        prefilledBlocks = new Block[size];
+        delegateIndexes = new int[size];
+
+        int outputIndex = 0;
+        int delegateIndex = 0;
+        for (HudiColumnHandle column : columns) {
+            if (partitionKeys.containsKey(column.getName())) {
+                String partitionValue = partitionKeys.get(column.getName());
+                Type type = column.getHiveType().getType(typeManager);
+                Object prefilledValue = deserializePartitionValue(type, partitionValue, column.getName(), timeZoneKey);
+                prefilledBlocks[outputIndex] = Utils.nativeValueToBlock(type, prefilledValue);
+                delegateIndexes[outputIndex] = -1;
+            }
+            else {
+                delegateIndexes[outputIndex] = delegateIndex;
+                delegateIndex++;
+            }
+            outputIndex++;
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegate.getCompletedBytes();
+    }
+
+    @Override
+    public long getCompletedPositions()
+    {
+        return delegate.getCompletedPositions();
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return delegate.getReadTimeNanos();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return delegate.isFinished();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        try {
+            Page dataPage = delegate.getNextPage();
+            if (dataPage == null) {
+                return null;
+            }
+            int batchSize = dataPage.getPositionCount();
+            Block[] blocks = new Block[prefilledBlocks.length];
+            for (int i = 0; i < prefilledBlocks.length; i++) {
+                if (prefilledBlocks[i] != null) {
+                    blocks[i] = new RunLengthEncodedBlock(prefilledBlocks[i], batchSize);
+                }
+                else {
+                    blocks[i] = dataPage.getBlock(delegateIndexes[i]);
+                }
+            }
+            return new Page(batchSize, blocks);
```

Comment:
This code though is right, is inefficient. This is one of the hot paths. If there is no prefilled blocks, short circuit and return the delegate page.

Reply:
Good catch, thank you. Add short circuit when no actual prefilled block.